primitives/tests/api.rs

```diff
@@ -0,0 +1,283 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! Test the API surface of `primitives`.
+//!
+//! The point of these tests are to check the API surface as opposed to test the API functionality.
+//!
+//! ref: <https://rust-lang.github.io/api-guidelines/about.html>
+
+#![allow(dead_code)]
+#![allow(unused_imports)]
+#![cfg(feature = "alloc")]
+#![cfg(feature = "arbitrary")]
+
+use arbitrary::Arbitrary;
+use bitcoin_primitives::block::{Checked, Unchecked};
+use bitcoin_primitives::script::{self, ScriptHash, WScriptHash};
+use bitcoin_primitives::{
+    absolute, block, merkle_tree, pow, relative, transaction, OutPoint, Script, ScriptBuf,
+    Sequence, Transaction, TxIn, TxOut, Txid, Witness, Wtxid,
+};
+use hashes::sha256t;
+
+/// A struct that includes all public non-error enums.
+#[derive(Debug)] // All public types implement Debug (C-DEBUG).
+struct Enums {
+    a: absolute::LockTime,
+    b: relative::LockTime,
+    // block::Checked, // Empty enums are not constructable.
+    // block::Unchecked,
+}
+
+/// A struct that includes all public non-error structs.
+#[derive(Debug)] // All public types implement Debug (C-DEBUG).
+struct Structs<'a> {
+    a: block::Block<Checked>,
+    b: block::Block<Unchecked>,
+    c: block::Header,
+    d: block::Version,
+    e: block::BlockHash,
+    f: block::WitnessCommitment,
+    g: merkle_tree::TxMerkleNode,
+    h: merkle_tree::WitnessMerkleNode,
+    i: pow::CompactTarget,
+    j: &'a Script,
+    k: ScriptHash,
+    l: WScriptHash,
+    m: ScriptBuf,
+    n: Sequence,
+    o: Transaction,
+    p: TxIn,
+    q: TxOut,
+    r: OutPoint,
+    s: Txid,
+    t: Wtxid,
+    u: transaction::Version,
+    v: Witness,
+    // w: witness::Iter<'a>,
+}
+
+static SCRIPT: ScriptBuf = ScriptBuf::new();
+static BYTES: [u8; 32] = [0x00; 32];
+
+/// Public structs that derive common traits.
+// C-COMMON-TRAITS excluding `Debug, Default, Display, Ord, PartialOrd, Hash`.
+#[derive(Clone, PartialEq, Eq)]
+struct CommonTraits {
+    a: block::Block<Checked>,
+    b: block::Block<Unchecked>,
+    c: block::Header,
+    d: block::Version,
+    e: block::BlockHash,
+    f: block::WitnessCommitment,
+    g: merkle_tree::TxMerkleNode,
+    h: merkle_tree::WitnessMerkleNode,
+    i: pow::CompactTarget,
+    // j: &'a Script,
+    k: ScriptHash,
+    l: WScriptHash,
+    m: ScriptBuf,
+    n: Sequence,
+    o: Transaction,
+    p: TxIn,
+    q: TxOut,
+    r: OutPoint,
+    s: Txid,
+    t: Wtxid,
+    u: transaction::Version,
+    v: Witness,
+    // w: witness::Iter<'a>,
```

Comment:
In d845272ddcf2ebe907a63eefaaeb504b59210f52:

We can do this in a followup, but IMO the iterator should implement `Clone` unless there's a good reason not to. `Debug`, `PartialEq` and `Eq` maybe not.

Reply:
IIRC `witness::Iter` is just a fancy counter so it indeed can impl `Clone`